BehaviorTreeEditor.cs

```diff
@@ -44,24 +61,120 @@ public static void ShowWindow()
 
 	void OnGUI()
 	{
+		try
+		{
+			CreateGUI();
+			AssetDatabase.SaveAssets();
+		}
+		catch ( Exception e )
+		{
+			Debug.LogError( e );
+		}
+	}
+
+	void CreateGUI()
+	{
 		CreateTypeLists();
 
-		_behaviorTree = (BehaviorTree)EditorGUILayout.ObjectField( "Behavior Tree",
-		                                                           _behaviorTree,
-		                                                           typeof( BehaviorTree ),
-		                                                           false );
+		_behaviorTree =
+		    (BehaviorTree)EditorGUILayout.ObjectField(
+		        "Behavior Tree",
+		        _behaviorTree,
+		        typeof( BehaviorTree ),
+		        false );
 
 		if ( _behaviorTree != null )
 		{
+			// set default root node
+			if ( !_behaviorTree.root )
+			{
+				_behaviorTree.root = nullNode;
+			}
+
 			GUILayout.Label( "Nodes", EditorStyles.boldLabel );
-			
-			Type resultType = CreateNodeTypeSelector( _behaviorTree.root );
+
+			Type resultType = CreateNodeTypeSelector( _behaviorTree.root, 2 );
 			if ( resultType != _behaviorTree.root.GetType() )
 			{
-				_behaviorTree.root = (TreeNode)Activator.CreateInstance( resultType );
+				DeleteNode( _behaviorTree.root );
+				_behaviorTree.root = CreateNode( resultType );
+			}
+			
+			CreateNodeGUI( _behaviorTree.root, 3 );
+		}
+	}
+
+	int CreateNodeGUI( TreeNode node, int height )
+	{
+		DebugUtils.Assert( node, "Cannot create GUI for null node!" );
+
+		++_indentLevel;
+		
+		if ( node is Decorator )
+		{
+			height = CreateDecoratorGUI( (Decorator)node, height );
+		}
+		else if ( node is Compositor )
+		{
+			height = CreateCompositorGUI( (Compositor)node, height );
+		}
+
+		--_indentLevel;
+
+		return height;
+	}
+
+	int CreateDecoratorGUI( Decorator node, int height )
+	{
+		if ( CreateChildrenFoldout( node, height ) )
+		{
+			Type resultType = CreateNodeTypeSelector( node._child, height );
```

Comment:
rename CreateNodeTypeSelector() to something like dropdown so it isn't confused with a selector node.